File: DSA_2215/queue/queue.dart
class Queue {
  List<int> _items = [];

  void add(int item) {
    _items.add(item);
  }

  int remove() {
    if (_items.isEmpty) {
      return -1;
    }
    return _items.removeAt(0);
  }

  void show() {
    for (var item in _items) {
      print("$item ");
    }
    print("");
  }

  int middle() {
    if (_items.isEmpty) {
      return -1;
    }
    int fast = 0;
    int slow = 0;
    while (fast < _items.length && fast + 1 < _items.length) {
      fast += 2;
      slow++;
    }
    return _items[slow];
  }

  bool search(int item) {
    return _items.contains(item);
  }
}

void useMe() {
  var queue = Queue();
  queue.add(1);
  queue.add(2);
  queue.add(3);
  queue.show();

  print("Middle element: ${queue.middle()}");

  print("Searching for 2: ${queue.search(2)}");

  var item = queue.remove();
  if (item != -1) {
    print("Removed item: $item");
  }

  queue.show();
}

void main() {
  useMe();
}

File: DSA_2215/queue/queue.go
package main

import (
	"container/list"
	"fmt"
)

type Queue struct {
	list *list.List
}

func NewQueue() *Queue {
	return &Queue{list: list.New()}
}

func (q *Queue) Add(item int) {
	q.list.PushBack(item)
}

func (q *Queue) Remove() int {
	element := q.list.Front()
	if element == nil {
		return -1
	}
	q.list.Remove(element)
	return element.Value.(int)
}

func (q *Queue) Show() {
	for e := q.list.Front(); e != nil; e = e.Next() {
		fmt.Printf("%d ", e.Value)
	}
	fmt.Println()
}

func (q *Queue) Middle() int {
	if q.list.Len() == 0 {
		return -1
	}
	fast := q.list.Front()
	slow := q.list.Front()
	for fast != nil && fast.Next() != nil {
		fast = fast.Next().Next()
		slow = slow.Next()
	}
	return slow.Value.(int)
}

func (q *Queue) Search(item int) bool {
	for e := q.list.Front(); e != nil; e = e.Next() {
		if e.Value == item {
			return true
		}
	}
	return false
}

func useMe() {
	queue := NewQueue()
	queue.Add(1)
	queue.Add(2)
	queue.Add(3)
	queue.Show()

	fmt.Printf("Middle element: %d\n", queue.Middle())

	fmt.Printf("Searching for 2: %t\n", queue.Search(2))

	item := queue.Remove()
	if item != -1 {
		fmt.Printf("Removed item: %d\n", item)
	}

	queue.Show()
}

func main() {
	useMe()
}
